### PR TITLE
8337980: Javac allows invocation of an inherited instance method from a static method

### PR DESCRIPTION
**Pull request by archiecobbs**
The method `Resolve.findFun()` performs the lookup for an unqualified method invocation like `foo()`. It does this by invoking `findMethod()` to find the method in each containing class scope, working from the inside out.

In this loop, as soon as it finds a matching method, but before returning, it performs the checks that prevent (a) invoking an instance method from a static context, and (b) invoking an instance method of the same class from an early-initialization context.

However, these checks don't account for the fact that in some cases `findMethod()` can return an `AmbiguityError` even though the code is perfectly valid. This is because `findMethod()` performs the "maximally specific" logic of JLS §15.12.2.5 ("Choosing the Most Specific Method"), but not the "preferred method" logic. Instead, in the case there are multiple "maximally specific" methods, they are all returned wrapped in an `AmbiguityError`; disambiguation is the caller's job.

When this happens, if the code is actually valid, there's no problem, but if the code has one of the errors (a) or (b), the error does not get detected, with resulting downstream chaos.

Here's an example of when `findMethod()` can return an `AmbiguityError` for perfectly valid code:
```java
public class Bug {

    public interface A {
        int op();
    }

    public abstract static class B {
        public abstract int op();
    }

    public abstract static class C extends B implements A {

        public int test() {
            return op();   // findMethod("op") returns both A.op() and B.op() here
        }
    }
}
```
The fix in this PR is to add a few lines to `findFun()` to resolve the ambiguity (if possible) before performing checks (a) and (b). It's possible there is an opportunity for a more thorough refactoring here, but short of that, I tried to add a few helpful comments.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8337980](https://bugs.openjdk.org/browse/JDK-8337980): Javac allows invocation of an inherited instance method from a static method (**Bug** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Contributors
 * Maurizio Cimadamore `<mcimadamore@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20533/head:pull/20533` \
`$ git checkout pull/20533`

Update a local copy of the PR: \
`$ git checkout pull/20533` \
`$ git pull https://git.openjdk.org/jdk.git pull/20533/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20533`

View PR using the GUI difftool: \
`$ git pr show -t 20533`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20533.diff">https://git.openjdk.org/jdk/pull/20533.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20533#issuecomment-2278739307)

### CODE REVIEW

**Comment by archiecobbs** [comment]
pingbot

File: test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        public abstract int op();
+    }
+
+    public abstract static class C extends B implements A {
+
+        public C(int x) {
+        }
+
+        public C() {
+            this(op());     // compile should fail here
```
Comment (mcimadamore) [inline test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java:23]: Question: the `mergeAbstracts` method IIRC returns a new method symbol whose owner and flags are those of the most specific signature. So, I wonder, if `B::op` returned `Object` and `A::op` returned `String`, would we still see the problem?

Comment (archiecobbs) [reply test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java:23]: By the time `mergeAbstracts` is called in the patched version of the code, the problem in the original version of the code has already occurred, i.e., the behavior of `mergeAbstracts` is not involved in how the bug happens. I verified that with those changes (`B::op` returning `Object` and `A::op` returning `String`) the verify error still occurs - even when you change the other constructor to `public C(String x)` (so that `B::op` would have an incompatible return value). I also verified that the modified example also works correctly with the PR applied.

Comment (mcimadamore) [reply test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java:23]: I guess what I'm asking is whether there's some way to alter the declaration of the problematic methods so that the early-constructor context error is not reported - this seems possible if mergeAbstracts merges info using the interface method as a base symbol?

Comment (mcimadamore) [reply test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java:23]: What I'm worried about is a situation where:
* multiple abstract methods are applicable
* one of the abstract method belongs to a class in an early construction context
* their signature needs to be merged
* when the signature is merged we use an interface as the owner for the symbol
* as a result, the early construction context error is missed (and I think the generated code will not pass the verifier)

Comment (archiecobbs) [reply test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java:23]: > one of the abstract method belongs to a class in an early construction context

I was thinking that because the two methods are inherited by a common subclass, and we are talking about an invocation `op()` in the context of that common subclass, then for any class `D` both method invocations are going to result in the same "class or interface to search" per §15.12.1 "Compile-Time Step 1: Determine Type to Search". And so then (the flexible constructors version of) §15.12.3 applies the same for both:

> Otherwise, let T be the class or interface to search (15.12.1). It is a compile-time error if ... the method invocation occurs in an early construction context (8.8.7.1) of class T...

In other words, it wouldn't be possible for one of the methods to "belong to a class in an early construction context
" and the other not to... which would imply your scenario can't happen, right? Can you come up with an example where that isn't true?

**Comment by mcimadamore** [comment]
I've double checked the code again, and I think I see where I made a wrong assumption. The "early construction" check in the code is based on the `env`. As such, it does not depend on whatever symbol is returned by `mergeAbstracts`. The code in this PR is ok.

Ultimately, this bug has to do with `findMethod` being called in a place (`findFun`) that does not follow up with the correct call to `mergeAbstract` (as this call is typically part of the `lookup` method in a `LookupHelper` - see e.g. `resolveMethod/resolveQualifiedMethod`).

What I'm a bit worried about with this fix is that it results in `mergeAbstracts` to be called *twice* when the code is correct. This seems bad, as `mergeAbstracts` is quite a complex routine. Ideally, `findFun` should return immediately if the result of `findMethod` "exists", and defer the check for static/early construction context to _after_ (e.g. when `mergeAbstracts` is normally called). Which I realize is challenging because by the time we call `mergeAbstract` outside `findFun`, we no longer have the `env` in which the symbol was found...

**Comment by mcimadamore** [comment]
Something like this seems to work ok:

```
$ git diff
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
index 5bf1bc0ead1..55293535ff8 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java
@@ -1853,6 +1853,10 @@ Symbol findMethod(Env<AttrContext> env,
                           bestSoFar,
                           allowBoxing,
                           useVarargs);
+        if (bestSoFar.kind == AMBIGUOUS) {
+            AmbiguityError a_err = (AmbiguityError)bestSoFar.baseSymbol();
+            bestSoFar = a_err.mergeAbstracts(site);
+        }
         return bestSoFar;
     }
     // where
@@ -2757,7 +2761,7 @@ Symbol resolveMethod(DiagnosticPosition pos,
         return lookupMethod(env, pos, env.enclClass.sym, resolveMethodCheck,
                 new BasicLookupHelper(name, env.enclClass.sym.type, argtypes, typeargtypes) {
                     @Override
-                    Symbol doLookup(Env<AttrContext> env, MethodResolutionPhase phase) {
+                    Symbol lookup(Env<AttrContext> env, MethodResolutionPhase phase) {
                         return findFun(env, name, argtypes, typeargtypes,
                                 phase.isBoxingRequired(),
                                 phase.isVarargsRequired());
@@ -2789,7 +2793,7 @@ private Symbol resolveQualifiedMethod(MethodResolutionContext resolveContext,
                                   List<Type> typeargtypes) {
         return lookupMethod(env, pos, location, resolveContext, new BasicLookupHelper(name, site, argtypes, typeargtypes) {
             @Override
-            Symbol doLookup(Env<AttrContext> env, MethodResolutionPhase phase) {
+            Symbol lookup(Env<AttrContext> env, MethodResolutionPhase phase) {
                 return findMethod(env, site, name, argtypes, typeargtypes,
                         phase.isBoxingRequired(),
                         phase.isVarargsRequired());
@@ -2913,7 +2917,7 @@ private Symbol resolveConstructor(MethodResolutionContext resolveContext,
                               List<Type> typeargtypes) {
         return lookupMethod(env, pos, site.tsym, resolveContext, new BasicLookupHelper(names.init, site, argtypes, typeargtypes) {
             @Override
-            Symbol doLookup(Env<AttrContext> env, MethodResolutionPhase phase) {
+            Symbol lookup(Env<AttrContext> env, MethodResolutionPhase phase) {
                 return findConstructor(pos, env, site, argtypes, typeargtypes,
                         phase.isBoxingRequired(),
                         phase.isVarargsRequired());
@@ -2972,7 +2976,7 @@ Symbol resolveDiamond(DiagnosticPosition pos,
         return lookupMethod(env, pos, site.tsym, resolveMethodCheck,
                 new BasicLookupHelper(names.init, site, argtypes, typeargtypes) {
                     @Override
-                    Symbol doLookup(Env<AttrContext> env, MethodResolutionPhase phase) {
+                    Symbol lookup(Env<AttrContext> env, MethodResolutionPhase phase) {
                         return findDiamond(pos, env, site, argtypes, typeargtypes,
                                 phase.isBoxingRequired(),
                                 phase.isVarargsRequired());
@@ -3503,18 +3507,6 @@ abstract class BasicLookupHelper extends LookupHelper {
             super(name, site, argtypes, typeargtypes, maxPhase);
         }
 
-        @Override
-        final Symbol lookup(Env<AttrContext> env, MethodResolutionPhase phase) {
-            Symbol sym = doLookup(env, phase);
-            if (sym.kind == AMBIGUOUS) {
-                AmbiguityError a_err = (AmbiguityError)sym.baseSymbol();
-                sym = a_err.mergeAbstracts(site);
-            }
-            return sym;
-        }
-
-        abstract Symbol doLookup(Env<AttrContext> env, MethodResolutionPhase phase);
-
         @Override
         Symbol access(Env<AttrContext> env, DiagnosticPosition pos, Symbol location, Symbol sym) {
             if (sym.kind.isResolutionError()) {
@@ -3561,10 +3553,6 @@ ReferenceLookupHelper unboundLookup(InferenceContext inferenceContext) {
         abstract JCMemberReference.ReferenceKind referenceKind(Symbol sym);
 
         Symbol access(Env<AttrContext> env, DiagnosticPosition pos, Symbol location, Symbol sym) {
-            if (sym.kind == AMBIGUOUS) {
-                AmbiguityError a_err = (AmbiguityError)sym.baseSymbol();
-                sym = a_err.mergeAbstracts(site);
-            }
             //skip error reporting
             return sym;
         }
```

(not saying you have to use this, mostly trying to explain more clearly what I meant with the above)

**Comment by archiecobbs** [comment]
> What I'm a bit worried about with this fix is that it results in `mergeAbstracts` to be called _twice_ when the code is correct.

OK now I understand your concern. This didn't occur to me before because I was assuming that `mergeAbstracts` did not modify any existing data. From a superficial look, that appears to be the case, but regardless as you might guess by my original comment that "It's possible there is an opportunity for a more thorough refactoring here" I like your patch much better. I was already wondering why the ambiguity resolution logic wasn't being incorporated into `findMethod()` (the method is named `findMethod()` not `almostFindMethod()` after all!) as you have done and assumed there was some mysterious reason for it. Apparently not. So let's go with your patch - thanks!

**Comment by mcimadamore** [comment]
> OK now I understand your concern. This didn't occur to me before because I was assuming that `mergeAbstracts` did not modify any existing data. From a superficial look, that appears to be the case

This is indeed the case - the method doesn't change existing data. So we can call it multiple times. But it's a complex method doing some complex analysis to scan all the applicable abstract symbols and merge them into a new method. Doing all that work just to throw it a way - and then doing it all over again is a bit sad (but we can obviously live with that).

**Review by mcimadamore — APPROVED** [review]

**Comment by mcimadamore** [comment]
@vicente-romero-oracle or @lahodaj please take a look. Thanks.

**Review by lahodaj — APPROVED** [review]
Looks good to me.

**Comment by archiecobbs** [comment]
Thanks for the reviews! I'm glad that it resulted in a nice code simplification as well.

/integrate
